### ISSUE TITLE
Does in-batch negatives contain negative passages from other in-batch questions?

### ISSUE THREAD

**Issue by wuyaoxuehun** [facebookresearch/DPR#141]
Does in-batch negatives regard all passages(both positive and negative) of other in-batch questions as negatives for a specific in-batch question(named Q)? Will these contain potential positive passages for Q?  I used to believe all positive passages( (instead of all passages) of other in-batch questions are regarded as in-batch negatives, But the dpr code base is against my thought.

**Comment by vlad-karpukhin** [comment]
"Does in-batch negatives regard all passages(both positive and negative) of other in-batch questions as negatives for a specific in-batch question(named Q)? " - yes, that is why they are called in-batch negatives.

"Will these contain potential positive passages for Q?" - in theory yes, in practice this is highly unlikely given a big and diverse train set. You can sure try to filter all negative passages in a batch so that they don't have answer spans for any of the questions. This will make training slower, and unlikely give any gains.
Though it may actually be helpful for datasets like SQuAD.

"I used to believe all positive passages( (instead of all passages) of other in-batch questions are regarded as in-batch negatives, But the dpr code base is against my thought." - this is correct understanding. DPR treats all other questions' positives (and negatives) as negatives for a given question.

What do you mean by exactly by "dpr code base is against my thought." ?

**Comment by wuyaoxuehun** [comment]
> 
> 
> "Does in-batch negatives regard all passages(both positive and negative) of other in-batch questions as negatives for a specific in-batch question(named Q)? " - yes, that is why they are called in-batch negatives.
> 
> "Will these contain potential positive passages for Q?" - in theory yes, in practice this is highly unlikely given a big and diverse train set. You can sure try to filter all negative passages in a batch so that they don't have answer spans for any of the questions. This will make training slower, and unlikely give any gains.
> Though it may actually be helpful for datasets like SQuAD.
> 
> "I used to believe all positive passages( (instead of all passages) of other in-batch questions are regarded as in-batch negatives, But the dpr code base is against my thought." - this is correct understanding. DPR treats all other questions' positives (and negatives) as negatives for a given question.
> 
> What do you mean by exactly by "dpr code base is against my thought." ?

Thanks for your prompt reply !
In your paper section 3.2-Positive and negative passages, you defined gold as "positive passages paired with other questions which appear in the training set", this is against what you did in your code, so I am a little confused.

**Comment by vlad-karpukhin** [comment]
sorry for a bit confused wording in the paper.

Term 'gold' is used in 2 contexts in the paper:

1. gold positive passages - passages provided by the datasets itself (but, in fact, their counterparts in our wikipedia split)

2. gold negatives - this is used just to describe that experiment with different types of negatives when calculating questions similarity vs a list of passages embeddings. 'Gold negatives' there are passages taken as gold positive passage but from different questions (so they are negatives for a given one). Such 'gold negatives' are implemented (along with hard negatives) in the current DPR's in-batch negatives loss calculation.

**Comment by wuyaoxuehun** [comment]
This makes it cleal! Thanks a lot!!
